services/shhext/service_test.go

```diff
@@ -231,6 +232,98 @@ func (s *ShhExtSuite) TestRequestMessages() {
 	s.Contains(api.service.tracker.cache, common.BytesToHash(hash))
 }
 
+func (s *ShhExtSuite) TestDebugPostSync() {
+	mock := newHandlerMock(1)
+	s.services[0].tracker.handler = mock
+	symID, err := s.whisper[0].GenerateSymKey()
+	s.NoError(err)
+	client, err := s.nodes[0].Attach()
+	s.NoError(err)
+	var hash common.Hash
+
+	var testCases = []struct {
+		name            string
+		hasPeers        bool
+		msg             whisper.NewMessage
+		postSyncTimeout time.Duration
+		expectedErr     error
+	}{
+		{
+			name:     "envelope expired",
+			hasPeers: false,
+			msg: whisper.NewMessage{
+				SymKeyID:  symID,
+				PowTarget: whisper.DefaultMinimumPoW,
+				PowTime:   200,
+				Topic:     whisper.TopicType{0x01, 0x01, 0x01, 0x01},
+				Payload:   []byte("hello"),
+				TTL:       1,
+			},
+			postSyncTimeout: postSyncTimeout,
+			expectedErr:     errEnvelopeExpired,
+		},
+		{
+			name:     "timeout",
+			hasPeers: true,
+			msg: whisper.NewMessage{
+				SymKeyID:  symID,
+				PowTarget: whisper.DefaultMinimumPoW,
+				PowTime:   200,
+				Topic:     whisper.TopicType{0x01, 0x01, 0x01, 0x01},
+				Payload:   []byte("hello"),
+			},
+			postSyncTimeout: postSyncTimeout,
+			expectedErr:     nil,
+		},
+		{
+			name:     "invalid message",
+			hasPeers: true,
+			msg: whisper.NewMessage{
+				PowTarget: whisper.DefaultMinimumPoW,
+				PowTime:   200,
+				Topic:     whisper.TopicType{0x01, 0x01, 0x01, 0x01},
+				Payload:   []byte("hello"),
+			},
+			postSyncTimeout: postSyncTimeout,
+			expectedErr:     whisper.ErrSymAsym,
+		},
+		{
+			name:     "context deadline exceeded",
+			hasPeers: true,
+			msg: whisper.NewMessage{
+				SymKeyID:  symID,
+				PowTarget: whisper.DefaultMinimumPoW,
+				PowTime:   10,
+				Topic:     whisper.TopicType{0x01, 0x01, 0x01, 0x01},
+				TTL:       100,
+				Payload:   []byte("hello"),
+			},
+			postSyncTimeout: 1 * time.Millisecond,
+			expectedErr:     errors.New("context deadline exceeded"),
+		},
+	}
+
+	for _, tc := range testCases {
+		s.T().Run(tc.name, func(t *testing.T) {
+			if tc.hasPeers {
+				s.nodes[0].Server().AddPeer(s.nodes[1].Server().Self())
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), tc.postSyncTimeout)
+			defer cancel()
+			err := client.CallContext(ctx, &hash, "debug_postSync", tc.msg)
+
+			if tc.expectedErr != nil {
+				s.Equal(tc.expectedErr.Error(), err.Error())
+			} else {
+				s.NoError(err)
+			}
+			if tc.hasPeers {
+				s.nodes[0].Server().RemovePeer(s.nodes[1].Server().Self())
```

Comment:
I would split test cases with and without peers into different test functions. The problem is that `Server.RemovePeer()` does not remove the peer immediately. It probably will fail if you put a test with `hasPeers: false` after a case with `hasPeers: true` or run these test cases in parallel.